### PR TITLE
DM-27843: Add anyFilterMapsToThis support to ReferenceObjectLoader

### PR DESCRIPTION
**Pull request by parejkoj**
As part of getting the gen3 code to support the more complicated filter mapping logic that the gen2 code did, I also refactored several methods into the common base class. Gen2 is going away soon, but we also want to keep as much of the gen2 test coverage as we can: this should help with that.

### CODE REVIEW

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 1]
```diff
@@ -461,6 +461,118 @@ def makeMinimalSchema(filterNameList, *, addCentroid=False,
...
+        with additional columns defined from the remaining function arguments.
+
+        Parameters
+        ----------
+        refCat : `lsst.afw.table.SimpleCatalog`
+            Reference catalog to map to new catalog
+        anyFilterMapsToThis : `str`
+            Always use this reference catalog filter.
+            Mutually exclusive with `filterMap`
+        filterMap : `dict` [`str`,`str`]
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/loadReferenceObjects.py:477]:
```suggestion
        filterMap : `dict` [`str`,`str`], optional
```

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 2]
```diff
@@ -461,6 +461,118 @@ def makeMinimalSchema(filterNameList, *, addCentroid=False,
...
+            Mapping of camera filter name: reference catalog filter name.
+        centroids : `bool`
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/loadReferenceObjects.py:479]:
```suggestion
        centroids : `bool`, optional
```

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 3]
```diff
@@ -461,6 +461,118 @@ def makeMinimalSchema(filterNameList, *, addCentroid=False,
...
+        or sets `anyFilterMapsToThis` in the schema.
+
+        Parameters
+        ----------
+        schema : `lsst.afw.table.Schema`
+            Schema for reference catalog.
+        anyFilterMapsToThis : `str`
+            Always use this reference catalog filter.
+            Mutually exclusive with `filterMap`
+        filterMap : `dict` [`str`,`str`]
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/loadReferenceObjects.py:527]:
```suggestion
        filterMap : `dict` [`str`,`str`], optional
```

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 4]
```diff
@@ -461,6 +461,118 @@ def makeMinimalSchema(filterNameList, *, addCentroid=False,
...
+        For each camFilter: refFilter in filterMap, adds these aliases:
+            <camFilter>_camFlux:      <refFilter>_flux
+            <camFilter>_camFluxErr: <refFilter>_fluxErr, if the latter exists
+        or sets `anyFilterMapsToThis` in the schema.
+
+        Parameters
+        ----------
+        schema : `lsst.afw.table.Schema`
+            Schema for reference catalog.
+        anyFilterMapsToThis : `str`
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/loadReferenceObjects.py:524]:
```suggestion
        anyFilterMapsToThis : `str`, optional
```

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 5]
```diff
@@ -461,6 +461,118 @@ def makeMinimalSchema(filterNameList, *, addCentroid=False,
...
+            Always use this reference catalog filter.
+            Mutually exclusive with `filterMap`
+        filterMap : `dict` [`str`,`str`]
+            Mapping of camera filter name: reference catalog filter name.
+            Mutually exclusive with `anyFilterMapsToThis`
+
+        Raises
+        ------
+        RuntimeError
+            If any required reference flux field is missing from the schema.
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/loadReferenceObjects.py:534]:
```suggestion
            Raised if any required reference flux field is missing from the schema.
```

**Review by TallJimbo — APPROVED** [review]
I reviewed commit-by-commit, and some code moved around, so it's possible the locations some changes should be made have moved.  But all comments are minor, and (I think) entirely about docs.

Overall, looks like a very nice cleanup.  I _think_ this doesn't actually touch the logic used to do the padding when actually loading refcat rows in regions, which is good - that code is _super_ fragile and not particularly well tested in CI, IIRC, and there are subtle differences between how Gen2 and Gen3 work that I think are important.  Someday we should try to fix that, of course, but for now I just want to make sure we avoid letting that problem torpedo unrelated "nearby" tickets like this one.

File: python/lsst/meas/algorithms/loadReferenceObjects.py [hunk 6]
```diff
@@ -295,35 +303,31 @@ def applyProperMotions(self, catalog, epoch):
...
-        config : `lsst.pex.config.configurableField`
-            Configuration for the loader.
-        log : `lsst.log.Log`, `logging.Logger` or `None`, optional
-            Logger object used to write out messages. If `None` a default
-            logger will be used.
-        """
+    Parameters
+    ----------
+    dataIds : iterable of `lsst.daf.butler.DataIds`
+        An iterable object of DataSetRefs which point to reference catalogs
```
[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/meas/algorithms/loadReferenceObjects.py]: I was hoping not to modify the logic of the gen3 loader code itself: this `__init__` gets called in potentially a lot of places (anything that uses a refcat and has a separate `runQuantum`, like jointcal and fgcmcal), so although that's probably a good cleanup, I think it needs its own ticket. It's also an API change, so might need an RFC, even if we all agree on it.

[the remark this answers is not in the document]
Comment (parejkoj) [reply python/lsst/meas/algorithms/loadReferenceObjects.py]: I filed this as https://jira.lsstcorp.org/browse/DM-33066

**Pull request merged**